### PR TITLE
Add support for multiple location sets

### PR DESCRIPTION
**Pull request by schrodit**
**What this PR does / why we need it**:
Added support for multiple locationSets.
These locations set bundle the currently used `testLocations` so that multiple locationSets (location profiles) can be defined.

To reference these locations, a paramter for a testflow step is introduced `location` where a specific locationSet can be defined.
If no reference is provided the default locationSet will be used.
This default locationSet can be specified with `default: true` if no locationSet is manually set to default, the first locationSet will be considered as default)

This change is backwards compatible which means that current testruns won't break and can be used in parallel. But the old way is now deprecated and the new sets should be used from now on.

**Special notes for your reviewer**:
any concerns about the naming?
Maybe rename `location` to `locationRef` or `locationSetRef`?

**Release note**:

```improvement user
Added support for multiple location sets.
```
```improvement user
Added support for specifying different versions for testflow steps.
```
```action user
`testLocations` are deprecated and will be removed in the future.
```

### CODE REVIEW

File: pkg/testmachinery/prepare/prepare.go [hunk 1]
```diff
@@ -83,11 +86,11 @@ func NewPrepare(name string, addOutput bool) (*PrepareDefinition, error) {
 	return prepare, nil
 }
 
-// AddLocation adds a tesdeflocation to the cloned repos and output artifacts.
-func (p *PrepareDefinition) AddLocation(loc Location) {
+// AddLocation adds a tesdef-location to the cloned repos and output artifacts.
```
Comment (dguendisch) [inline pkg/testmachinery/prepare/prepare.go]:
```suggestion
// AddLocation adds a testdef-location to the cloned repos and output artifacts.
```

**Comment by dguendisch** [comment]
Naming: I'd prefer `locationSet` so that the fields are named the same. `locationSetRef` would probably hint me that we are referring to another k8s resource here, so I wouldn't add Ref.

File: pkg/testmachinery/locations/locationsets.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+// NewLocations returns locations interface for a testrun
+func NewLocations(spec tmv1beta1.TestrunSpec) (Locations, error) {
+	if spec.LocationSets != nil {
+		return NewSetLocations(spec.LocationSets)
+	}
+
+	if len(spec.TestLocations) > 0 {
+		return NewTestLocation(spec.TestLocations)
```
[the remark this answers is not in the document]
Comment (schrodit) [reply pkg/testmachinery/locations/locationsets.go]: Sure your're wright. The function also returns testLocations type

[the remark this answers is not in the document]
Comment (schrodit) [reply pkg/testmachinery/locations/locationsets.go]: updated

**Pull request merged**